prime-router/src/main/kotlin/azure/ActionHistory.kt

```diff
@@ -713,4 +756,194 @@ class ActionHistory {
+                        // Add 2 to account for array offset and csv header
+                        rowsByGroupingId[groupingId]?.add(resultDetail.row + 1)
+                    }
+                }
+                it.writeArrayFieldStart(field)
+                rowsByGroupingId.keys.forEach { groupingId ->
+                    it.writeStartObject()
+                    it.writeStringField("message", messageByGroupingId[groupingId])
+                    it.writeStringField("rows", createRowsDescription(rowsByGroupingId[groupingId]))
+                    it.writeEndObject()
+                }
+                it.writeEndArray()
+            }
+            writeConsolidatedArray("consolidatedErrors", errors)
+            writeConsolidatedArray("consolidatedWarnings", warnings)
+        }
+        return outStream.toString()
+    }
+
+    /**
+     * Creates a list of ItemRouting instances with the report index
+     * and trackingId along with the list of the receiver organizations where
+     * the report was routed.
+     * @param report the instance generated while processing the report
+     * @param actionHistory the instance generated while processing the report
+     * @return the report routing for each item
+     */
+    private fun createItemRouting(
+        report: Report?,
+    ): List<ItemRouting> {
+        // create the item routing from the item lineage
+        val routingMap = mutableMapOf<Int, ItemRouting>()
+        this.itemLineages.forEach { il ->
+            val item = routingMap.getOrPut(il.parentIndex) { ItemRouting(il.parentIndex, il.trackingId) }
+            this.reportsOut[il.childReportId]?.let { rf ->
+                item.destinations.add("${rf.receivingOrg}.${rf.receivingOrgSvc}")
+            }
+        }
+        // account for any items that routed no where and were not in the item lineage
+        return report?.let {
+            val items = mutableListOf<ItemRouting>()
+            // the report has all the submitted items
+            it.itemIndices.forEach { i ->
+                // if an item was not present, create the routing with empty destinations
+                items.add(
+                    routingMap.getOrDefault(
+                        i,
+                        ItemRouting(i, it.getString(i, it.schema.trackingElement ?: ""))
+                    )
+                )
+            }
+            items
+        } ?: run {
+            // unlikely, but in case the report is null...
+            routingMap.toSortedMap().values.map { it }
+        }
+    }
+
+    /**
+     * Saves covid result metadata to the database
+     */
+    private fun insertCovidMetadata(
+        txn: Configuration
+    ) {
+        covidMetaDataRecords.forEach {
+            context?.logger?.info("Writing deidentified report data to the DB")
+            // wrap the insert into an exception handler
+            try {
+                DSL.using(txn).newRecord(COVID_RESULT_METADATA, it).store()
```

Comment:
This behavior used to use the `saveTestData` function from `DatabaseAccess.kt`. It looks like that has different behavior, is what's here equivalent or better?

Reply:
Updated